### PR TITLE
Improve docs

### PR DESCRIPTION
**Pull request by janEbert**
If you disagree with any of the changes, please tell me what to reverse or fix.
I am unsure about the docstrings I added to `src/utils.jl` for `unsqueeze` and
the `[un]stack` functions so please give those a more detailed look.

Update Documenter.jl version for new features, fix deprecation warnings in
`docs/make.jl` and import Flux for all doctests.
Add missing docstrings to `src/utils.jl`, `src/layers/stateless.jl` and `src/data/`; add
these and other missing functions to Markdown docs.

Improve docstrings by...
   - fixing typos,
   - removing trailing or double whitespaces,
   - using `jldoctest` blocks where applicable,
   - fixing, updating or correctly setting up existing doctests,
   - improving consistency (for example, always use "# Examples" instead
     of other variants),
   - removing empty lines between docstrings and functions,
   - instead of mentioning keywords, put them into the docstring,
   - adding some missing but useful keywords,
   - adding references (`@ref`),
   - using LaTeX math where applicable, and
   - linking papers.

Debatable stuff that is untouched:
   - BE/AE s/z irregularities (e.g. "normalise" versus "normalize") since
     most papers use the AE version while the Flux source code was
     written with BE spelling.
   - Names of normalization functions are capitalized
     ("Batch Normalization" instead of "batch normalization").
   - Default values in argument lists have spaces around the equals sign (`arg = x` instead of `arg=x`).

### CODE REVIEW

**Comment by janEbert** [comment]
Travis fails due to an old (i.e. incompatible) Documenter.jl version in `docs/Manifest.toml`. I changed the version to be the most recent one.

**Comment by janEbert** [comment]
Sorry, I accidentally included your commits in a rebase and now I am listed as the comitter. How do I revert this?

**Comment by MikeInnes** [comment]
You probably want something like `git rebase -i master`.

**Comment by janEbert** [comment]
Thanks a bunch, that did the trick!

File: src/layers/stateless.jl [hunk 1]
```diff
@@ -2,47 +2,89 @@ using NNlib: logsoftmax, logσ
...
+    crossentropy(ŷ, y; weight = 1)
+
+Return the cross entropy between the given probability distributions.
+
+# Examples
+```jldoctest
+julia> Flux.crossentropy(softmax([-1.1491, 0.8619, 0.3127]), [1, 1, 0])
+3.085467254747739
+```
+"""
```
[the remark this answers is not in the document]
Comment (janEbert) [reply src/layers/stateless.jl]: Thanks, that's perfect!

**Comment by janEbert** [comment]
I have made a [new branch](https://github.com/janEbert/Flux.jl/tree/doc-patch-rebased) with the changes rebased on top of the [current master](https://github.com/FluxML/Flux.jl/tree/0d3aa8fa5e03085d55936d2f123187f0e3ff7edd).
I haven't given everything a detailed look yet, though. I can either force push the new, rebased branch, discarding this one or close this PR and make a new one with the new branch.

**Comment by MikeInnes** [comment]
I think force pushing to this branch would be fine. Thanks a lot for looking at this (and sorry for the churn) – looking forward to having this.

**Comment by janEbert** [comment]
No worries! I'm very happy to give something back. :)
I just looked through the changes again, those seem to be fine. However, some doctests fail due to the Zygote changes. After fixing those, this should be ready to merge!

**Comment by dhairyagandhi96** [comment]
Could we do a rebase

**Comment by janEbert** [comment]
Rebase done.
I also changed some stuff from #875, mostly redundancies like listing defaults twice. I'd also suggest not explicitly documenting every parameter:
```
    WeightDecay(wd = 0)

Decay weights by `wd`.

# Parameters
  - Weight decay (`wd`)
```
Kind of comedic, isn't it? :D
This really does not seem worth making a whole extra section for. I didn't touch these though. If you want, I can remove parameter lists like the above.

Also, most examples don't really clarify or exemplify anything – they just construct the object with the default or non-default parameters. I didn't touch these either but I would strongly suggest pondering whether these are worth keeping.

**Comment by dhairyagandhi96** [comment]
There was some discussion in the past not to have the parameter defaults being a part of the API (causing confusion with being taken for kwargs), any reason to go back to that?

 In the same vein, any example may show creating an optimiser. The update steps will be the same for all of them. It would of course be really helpful to have something that achieves this and can reduce bloat at the same time. What are your thoughts?

**Comment by janEbert** [comment]
If they aren't part of the API, then they shouldn't be documented either, or am I not understanding correctly?
The signature string is just plain Julia syntax and it's the [recommended way](https://docs.julialang.org/en/v1/manual/documentation/) (relevant part quoted below) – I don't see a reason to write it differently. Keyword arguments are indicated by a semicolon before the list so there is no possibility of confusion.
I had also applied the same changes to other docstrings (for example in `layers/conv.jl`) which wasn't complained about (yet ;) ). Doing it differently for the optimisers would be inconsistent.

By the way, I haven't touched signatures with `arg = 1` either. It seemed like the spaces around the equal sign were desired even though it's recommended to omit the spaces.

Here the part about documenting function signatures:
> [...] Optional arguments should be represented with their default values (i.e. f(x, y=1)) when possible, following the actual Julia syntax. Optional arguments which do not have a default value should be put in brackets (i.e. f(x[, y]) and f(x[, y[, z]])). An alternative solution is to use several lines: one without optional arguments, the other(s) with them. This solution can also be used to document several related methods of a given function. When a function accepts many keyword arguments, only include a <keyword arguments> placeholder in the signature (i.e. f(x; <keyword arguments>)), and give the complete list under an # Arguments section (see point 4 below).

**Comment by janEbert** [comment]
> In the same vein, any example may show creating an optimiser. The update steps will be the same for all of them. It would of course be really helpful to have something that achieves this and can reduce bloat at the same time. What are your thoughts?

Maybe just reference the example from `Descent`? That's the only "complete" one anyway.

**Comment by janEbert** [comment]
Hey, I really didn't mean to be rude with my message and by pushing the commits dispersing even further from the original documentation you wrote. I simply tried implementing defaults the Julia team has decided on so everything is nice, consistent and easily extensible by following the official documentation.
If you think the way I did it is wrong, then I can revert those changes (I just want to improve the newcomer experience) but I personally am strongly in favor of doing it the officially sanctioned way.

Also, what do you think about referencing `Descent` as an example application of an optimizer? Instead of listing the same example for all optimizers, this would reduce bloat.

**Comment by MikeInnes** [comment]
I don't think any rudeness was taken, it's just taking some time to cycle through all the PRs right now (thanks a lot for your patience here).

On default arguments: The problem with that style is that it's confusing for people not familiar with how default and kwarg arguments are defined in Julia (most users, as we grow). It reflects the method's _definition_, but not how the method is actually _called_ by users, which a problem; if you actually paste `foo(x, y=1)` into your repl, where `y` is a default positional argument, it'll throw an error. I think the docs are mistaken here (understandably, as we made the mistake also) and should be changed.

**Comment by janEbert** [comment]
Yeah, that behaviour is one big issue in Julia; it would be great for self-documenting code to be able to write positional arguments this way, too, like in Python (although without changing the positions).
I see your point, but shouldn't this rather be fixed in Julia itself instead of documenting around it? There was also some discussion about documentation related to default positional arguments  [here](https://github.com/JuliaLang/julia/issues/4902) (even your argument of not being able to paste the code was brought up).

A very ugly solution to keep the concise syntax would be to comment positional arguments similar to how IntelliJ does it.
`foo(x, y=1; z=0)` could become `foo(x, #=y=# 1; z=0)`.
Again, super ugly but it would work, though giving users the wrong impression.
The other, more obvious, solution would be to list default positional arguments for _all_ functions, not just for optimizers if you really wanted to go that way. Either way, the optimizer documentation is currently inconsistent with other modules.

(By the way, is there an issue for allowing specifying positional arguments by name? I've sought around a lot but never found one even though it seems obvious to ask for it. It would be great if it was possible, even if the name of the argument was ignored (due to issues with multiple dispatch).)

**Comment by MikeInnes** [comment]
> shouldn't this rather be fixed in Julia itself instead of documenting around it? 

Much as I like to be optimistic, the Flux documentation has to reflect how the Julia language is now, not how we'd like it to be. If the language changes, we'll make updates as necessary.

> Either way, the optimizer documentation is currently inconsistent with other modules.

Lucky for us, humans are pretty flexible, so it's fine to give up consistency in all cases for clarity in specific cases. One of the reasons that optimisers are a special case is that the (inherited) parameters names `eta`, `rho` etc are pretty undescriptive, so we need to document them clearly anyway; throwing in default values at the same time seems like a neat solution. If other modules would benefit from parameter lists, we'll add them there too, but consistency is not a goal in itself.

**Comment by janEbert** [comment]
It seems you don't want to have this discussion but I don't really see why the docs are responsible for fixing badly named variables. Also, your argument was that the docs should give newcomers an easier time by not listing default positional arguments the official way. Then wouldn't other modules profit from this change all the same?

I feel your opinion is pretty clear, though; I'm going to change the optimizers back to how they were.

**Comment by janEbert** [comment]
Well, it's like calling a parameter `mu` instead of `mean`. You are really just writing down the abstract, arbitrary, mathematical symbol instead of documenting (in your code) the _meaning_ of that abstract symbol. And I don't think I've ever seen a paper talk about "stochastic gradient descent with rho". :)
While they still call `Adam`'s momentum decay "`betas`", the largest deep learning frameworks all use intuitive names instead of symbols used in formulas for their arguments. TensorFlow, PyTorch and MXNet all call it `learning_rate` or `lr` instead of `alpha` or `eta`. They also say `momentum` instead of `rho`. Although even those frameworks aren't that consistent in naming parameters meaningfully instead of abstractly, it's simply much more digestible. I'd be glad to open another PR to change the variable names (or maybe an issue first to get some opinions?).
We could also reference the names used in mathematical formulae in the documentation instead of the other way around as it is currently. For example: 
```
`eta`: learning rate
# would become something like
`learning_rate`: [...] ``\eta`` in the literature.
```
While the documentation in this "direction" is more awkward, it's often times not even necessary as the meaning of the parameter should be clear intuitively.

**Comment by MikeInnes** [comment]
Yes, let's have a new issue that goes through everything, and discuss it a bit. ADAM's parameters seem harder but we can probably figure something out. With good ascii names, it may also make sense to just make all of these things kwargs anyway, which would help with the documentation issue.

**Comment by dhairyagandhi96** [comment]
@janEbert absolutely no rudeness taken! In fact, I'm pretty grateful that such engaged discussion happens on these matters since these are quite literally how newcomers would interact with Flux. Thanks for all the excellent work too! I'd taken a few personal days off to spend some time with fam, sorry about the delay.

A kwarg-only constructor would be fine too, if we are okay with the obvious code repitition compromise.

**Comment by janEbert** [comment]
I'm glad about that, and thank you! Hope you had a good time with the family. :)

**Comment by aminya** [comment]
I made a PR to this PR here: https://github.com/janEbert/Flux.jl/pull/1

**Comment by aminya** [comment]
What else should be done to merge this?

File: src/data/cmudict.jl [hunk 2]
```diff
@@ -24,18 +24,35 @@ function load()
   end
 end
 
+"""
+    phones()
+
+Return a `Vector` containing the phones used in the dataset.
```
[the remark this answers is not in the document]
Comment (janEbert) [reply src/data/cmudict.jl]: Sounds good, I'll change it. Also sorry I haven't reverted the parameter stuff yet, will get to that sometime soon.

**Comment by aminya** [comment]
I want to improve documentation, but because this is not on Master, I can't do it. Can you please merge this (or part of the work which everyone agrees with)?

**Comment by aminya** [comment]
This PR stayed on hold for several months, and because of that, now it has many conflicts. 😞

**Comment by CarloLucibello** [comment]
many doc improves here have been done in other PRs, but there is some stuff that left that we could use

**Comment by aminya** [comment]
> 
> 
> many doc improves here have been done in other PRs, but there is some stuff that left that we could use

Could you fix the conflicts? Since you have access to this repository, you should be able to do it.

**Comment by CarloLucibello** [comment]
to resolve this many conflicts takes a lot of time I don't have right now

**Comment by janEbert** [comment]
Hey hey! So sorry for never getting back to this. I'll fix the conflicts and revert the undesired changes today or tomorrow and hopefully the rest can be merged as is!

I haven't worked with Flux for some months now so if a lot has changed, it may take me an extra day to read through everything again and correct new docs (although I don't think it will take that long).

**Comment by janEbert** [comment]
Thanks aminya for bumping!

**Comment by janEbert** [comment]
Hey! Everything is rebased onto master and all good now; I even fixed some new docstrings and some other problems I found (for example the "missing docstring" message [here (current stable)](https://fluxml.ai/Flux.jl/v0.10.3/models/basics/#Utility-functions-1) or unlisted docstrings like for [RADAM](https://fluxml.ai/Flux.jl/v0.10.3/training/optimisers/#Optimiser-Reference-1)).
I also reverted the discussed, undesired changes to optimiser parameters as I said I would.

If you have any more small corrections, please tell me. I really want this to be merged soon, but see the time I'm willing to put into this coming to an end.
However, if these changes are simply not desired, I understand and will close the PR. I have noticed a lack of love for the docs (see the missing docstring in the stable docs, countless redundant whitespaces, and language errors continually being introduced).

This lack of attention to detail does not increase my motivation to continue this PR any further. However, since this seems to be okay for you, just go on with the current way! It's definitely easier that way and I really don't judge. :)
But _do_ merge this before going back to inserting spaces at line endings! It really is not that much work to skip through [this list](https://github.com/FluxML/Flux.jl/pull/853/files) since most of it is language/symbol corrections.

Cheers and sorry I took so long. I hope this is not taken the wrong way.

**Comment by janEbert** [comment]
Do the doctests for `chunk` fail due to a different CI Flux version? When I execute the following in my Flux dev repo I get the expected output (on both the PR head and your master):
```julia
$ julia --project
julia> using Flux

julia> Flux.chunk(1:10, 3)
3-element Array{UnitRange{Int64},1}:
 1:4
 5:8
 9:10
```

File: docs/src/training/training.md [hunk 3]
```diff
@@ -15,7 +15,7 @@ Flux.Optimise.train!
 
 There are plenty of examples in the [model zoo](https://github.com/FluxML/model-zoo).
 
-## Loss Functions
+## Loss
```
Comment (CarloLucibello) [inline docs/src/training/training.md:18]: I prefer Loss Functions here

**Comment by CarloLucibello** [comment]
Thanks for this, very nice work. 
CI is currently failing due to a problem in Boston Housing download, I'm fixing it now. 
I'm also activating CI for julia 1.4. Can we do doctesting only on a specific julia version?

File: docs/src/training/training.md (hunk 3 above)
Comment (janEbert) [reply docs/src/training/training.md:18]: Absolutely; that's an artifact from when I added documentation for loss functions. It has since been added elsewhere.
Thanks for the feedback and checking through!

**Comment by janEbert** [comment]
I tried to look for it in the Documenter.jl manual but couldn't find anything. Seems like it is currently not possible to do [versioned doctests](https://github.com/JuliaDocs/Documenter.jl/issues/1050).

A suggestion in the issue is actually already being used for the runtests.jl doctests. Currently doctests are run only for `VERSION >= v"1.2"`; we could apply the same for docs/make.jl.

**Comment by CarloLucibello** [comment]
all right, let's merge this. Thank you for the perseverance!

bors r+

**Comment by janEbert** [comment]
Thank _you_ for the quick response! I'm happy to see it coming into fruition.
I can still squash and whatnot if you want.

**Comment by aminya** [comment]
Finally 🚀
